### DISCUSSION TITLE
Hey Marvin

### DISCUSSION THREAD

**Discussion by StuartIanNaylor** [dscripka/openWakeWord#6] [General]
Would you train a Hey-Marvin model so that I can run a tests on a collection of non synthetic even if concatenated 'hey marvin' KW.

I am struggling to get much of a dataset for any of the KW you have and you would think with Mycroft being open source there would be an open dataset but not as I know...

**Comment by dscripka** [comment]
Sure thing, I'll start generating the training data for a "hey marvin" model.

I agree that it is difficult to find good keyword datasets. The Picovoice Wakeword benchmark dataset is very useful (https://github.com/Picovoice/wake-word-benchmark/tree/master/audio), but it only has several hundred examples per wake word. Another good option is Mozilla's single word target segment dataset (https://commonvoice.mozilla.org/en/datasets) which has a relatively large number of "firefox" clips. I haven't trained a "firefox" model yet, but that may also be a good next step.

**Comment by StuartIanNaylor** [comment]
The single word target segment is where I get 'Hey' from 'Marvin' is from the Google command set where I have some routines for trimming, concatenating, augmenting and adding background to in ProjectEars in the Dataset section.

I am not sure about the validity of the Picovoice Wakeword benchmark dataset for comparison to be honest as even when creating KW myself due to the huge amount of dross alignment and recording I end up using a model to filter out bad.
I will stand by my false positives benches, but head scratching how to combat cherry picking KW datasets that fit a model than vice versa.

I like how you are synthetically creating KW with different voices and it would be really injteresting to see what the Picovoice results are with a synthetic dataset that is not cherry picked for false postives.

**Comment by dscripka** [comment]
Ah, I see, that's a good usage of those datasets. I did something similar in the past, collecting some wakeword examples from youtube videos and podcasts by searching through aligned transcripts. It worked fairly well for that word, but it is effort intensive and difficult to scale (at least with how I was doing it).

I agree that it is difficult to properly evaluate wake word models as the best data comes from actual usage and not benchmark datasets. The Picovoice data (and other similar datasets) are useful for at least showing baseline performance, but ultimately you just have to try the models in a deployment scenario to see if they work well enough.

Testing for false positives is a little bit easier, but is still challenging. One of the reasons I chose the [DIPCO dataset](https://www.amazon.science/publications/dipco-dinner-party-corpus) is that it's reasonably close to a deployment environment: far-field audio of conversational speech, the presence of background noise & music, etc. But I'm interested to hear bout the methods you use to estimate the false-positive rate.

With respect to the "hey marvin" model, I pushed a new commit to the repo with a trained model for that wake word: https://github.com/dscripka/openWakeWord/blob/main/openwakeword/resources/models/hey_marvin_v0.1.onnx

From my testing on your "hey marvin" dataset (at least the version I found), it looks like the openWakeWord model gets around a ~2.6% false reject rate at a false accept rate of 0.5 per hour.

![hey_marvin_roc_curve](https://user-images.githubusercontent.com/9630975/210194033-ef699399-4611-434e-9ee1-8482dc5ecd0c.png)

I also pushed a notebook that shows the methods I used to calculate these metrics: https://github.com/dscripka/openWakeWord/blob/main/notebooks/performance_metrics.ipynb

**Reply by StuartIanNaylor** [reply]
Not very pythonic but I just hacked 2 seperate with 100 hours of speech aka https://www.openslr.org/resources/12/train-clean-100.tar.gz

```
import tensorflow as tf
import numpy as np
import glob
import os
import soundfile as sf

def softmax_stable(x):
    return(np.exp(x - np.max(x)) / np.exp(x - np.max(x)).sum())   

def kw_detect(rec, sample_rate ,duration, reset_state):


    rec = np.reshape(rec, (1, int(sample_rate * duration)))
    #rec = np.multiply(rec, 8)
    if reset_state:
      for s in range(len(input_details1)):
        inputs1[s] = np.zeros(input_details1[s]['shape'], dtype=np.float32)
    # Make prediction from model
    interpreter1.set_tensor(input_details1[0]['index'], rec)
    # set input states (index 1...)
    for s in range(1, len(input_details1)):
      interpreter1.set_tensor(input_details1[s]['index'], inputs1[s])
  
    interpreter1.invoke()
    output_data = interpreter1.get_tensor(output_details1[0]['index'])
    # get output states and set it back to input states
    # which will be fed in the next inference cycle
    for s in range(1, len(input_details1)):
      # The function `get_tensor()` returns a copy of the tensor data.
      # Use `tensor()` in order to get a pointer to the tensor.
      inputs1[s] = interpreter1.get_tensor(output_details1[s]['index'])
       
    out_softmax = softmax_stable(output_data[0])
    return out_softmax[0]

# Parameters
duration = 0.020
sample_rate = 16000
num_channels = 1

# Load the TFLite model and allocate tensors.
interpreter1 = tf.lite.Interpreter(model_path="../GoogleKWS/models2/crnn_stateb3/quantize_opt_for_size_tflite_stream_state_external/stream_state_external.tflite", num_threads=2)

interpreter1.allocate_tensors()

# Get input and output tensors.
input_details1 = interpreter1.get_input_details()
output_details1 = interpreter1.get_output_details()

inputs1 = []

for s in range(len(input_details1)):
  inputs1.append(np.zeros(input_details1[s]['shape'], dtype=np.float32))
kw_hit_qty = 0
total_duration = 0.0
hit_txt = []
reset_state = True
kw_hit_rbuff = np.zeros(10, dtype=np.float32)
reset_rec =np.zeros(320, dtype=np.float32)
for txtfile in glob.glob('/media/stuart/New Volume/Users/Stuart/Downloads/Noise/LibriSpeech/**/*.txt', recursive=True):
    dirtxt = os.path.dirname(txtfile)
    with open(txtfile) as f:
      lines = f.readlines()
      for line in lines:
       frame = 0
       kw_count = 0
       kw_hit = False
       content = line.split(" ", 1)
       flacfile = dirtxt + '/' + content[0] + '.flac'
       data, samplerate = sf.read(flacfile, dtype='float32')
       max_amp = np.max(np.abs(data))
       data = np.multiply(data, 0.8 / max_amp)
       total_duration = total_duration + (len(data) / samplerate)
       max_amp = np.max(np.abs(data))
       while frame < int(len(data) / 320):
         start = 320 * frame
         rec = data[start:start + 320]
         if len(rec) < 320:
           break
         kw_prob = kw_detect(rec, sample_rate ,duration, False)
         kw_hit_rbuff = np.roll(kw_hit_rbuff, -1)
         kw_hit_rbuff[len(kw_hit_rbuff) - 1] = kw_prob
         kw_hit_prob = np.mean(kw_hit_rbuff)
         if kw_prob > 0.99995:
           kw_hit = True
           kw_prob = kw_detect(reset_rec, sample_rate ,duration, True)
           kw_hit_rbuff = np.zeros(10, dtype=np.float32)
           print(flacfile, kw_prob, frame)
         frame += 1
       if kw_hit == True:
         kw_hit_qty += 1
         hit_txt.append(flacfile)
       print(kw_hit_qty, total_duration / 3600)
       
print(kw_hit_qty, total_duration / 3600)
```
Then I used https://zenodo.org/record/2529934#.Y6JLqNLP2RQ for noise 50 hours
```
import tensorflow as tf
import numpy as np
import glob
import os
import soundfile as sf

def softmax_stable(x):
    return(np.exp(x - np.max(x)) / np.exp(x - np.max(x)).sum())   

def kw_detect(rec, sample_rate ,duration, reset_state):


    rec = np.reshape(rec, (1, int(sample_rate * duration)))
    #rec = np.multiply(rec, 8)
    if reset_state:
      for s in range(len(input_details1)):
        inputs1[s] = np.zeros(input_details1[s]['shape'], dtype=np.float32)
    # Make prediction from model
    interpreter1.set_tensor(input_details1[0]['index'], rec)
    # set input states (index 1...)
    for s in range(1, len(input_details1)):
      interpreter1.set_tensor(input_details1[s]['index'], inputs1[s])
  
    interpreter1.invoke()
    output_data = interpreter1.get_tensor(output_details1[0]['index'])
    # get output states and set it back to input states
    # which will be fed in the next inference cycle
    for s in range(1, len(input_details1)):
      # The function `get_tensor()` returns a copy of the tensor data.
      # Use `tensor()` in order to get a pointer to the tensor.
      inputs1[s] = interpreter1.get_tensor(output_details1[s]['index'])
       
    out_softmax = softmax_stable(output_data[0])
    return out_softmax[0]

# Parameters
duration = 0.020
sample_rate = 16000
num_channels = 1

# Load the TFLite model and allocate tensors.
interpreter1 = tf.lite.Interpreter(model_path="../GoogleKWS/models2/crnn_stateb3/quantize_opt_for_size_tflite_stream_state_external/stream_state_external.tflite")

interpreter1.allocate_tensors()

# Get input and output tensors.
input_details1 = interpreter1.get_input_details()
output_details1 = interpreter1.get_output_details()

inputs1 = []

for s in range(len(input_details1)):
  inputs1.append(np.zeros(input_details1[s]['shape'], dtype=np.float32))
kw_hit_qty = 0
total_duration = 0.0
hit_txt = []
reset_state = True
kw_hit_rbuff = np.zeros(10, dtype=np.float32)
reset_rec =np.zeros(320, dtype=np.float32)
for flacfile in glob.glob('/media/stuart/New Volume/Users/Stuart/Downloads/Noise/FSDnoisy18k.audio_train/*.wav'):
       frame = 0
       kw_count = 0
       kw_hit = False
       data, samplerate = sf.read(flacfile, dtype='float32')
       max_amp = np.max(np.abs(data))
       data = np.multiply(data, 0.8 / max_amp)
       total_duration = total_duration + (len(data) / samplerate)
       while frame < int(len(data) / 320):
         start = 320 * frame
         rec = data[start:start + 320]
         if len(rec) < 320:
           break
         kw_prob = kw_detect(rec, sample_rate ,duration, False)
         kw_hit_rbuff = np.roll(kw_hit_rbuff, -1)
         kw_hit_rbuff[len(kw_hit_rbuff) - 1] = kw_prob
         kw_hit_prob = np.mean(kw_hit_rbuff)
         if kw_prob > 0.99995:
           kw_hit = True
           kw_prob = kw_detect(reset_rec, sample_rate ,duration, True)
           kw_hit_rbuff = np.zeros(10, dtype=np.float32)
           print(flacfile, kw_prob, frame)
         frame += 1
       if kw_hit == True:
         kw_hit_qty += 1
         hit_txt.append(flacfile)
       print(kw_hit_qty, total_duration / 3600)
       
print(kw_hit_qty, total_duration / 3600)
```

What I am liking is your synthetic KW dataset creation as when I am mining GSC or ML-commons I have to use models to filter the dataset as an example some of the 'Hey' actually say 'Hey DJ' and others are trimmed and padded completely wrong whilst others just contain mic hum.
With the synthetic KW test dataset at least we have a dataset that is free from curration intenional or not and also does the later Nvidia TTS not also have emotional variance?

I use Librispeech Clean 100 as its set in stone as if I augment I add a random element that isn't comparitive, but what we could do is share a test dataset that has been created synthetically even though models trained on that synthetic fingerprint will obviously have a head start.
I can start adding synthetic data to my dataset which another thing that I do as often KWS are binary classification models of KW and  Unknown there is so little cross entropy the model instantly overfits and unknown alone is so wide its unamanageable.
Much was garnered from https://github.com/google-research/google-research/tree/master/kws_streaming that also has a Noise classification so that Unkown becomes !KW voice and Noise is !voice that I have my own collection for.

I get zero false positives on 50 hours of noise as false positives occur only with the 100 hours of speech and I also add soundslike classification to deliberately create cross entropy around the KW which forces the model to train harder and focus on the entropy around the KW but it can make dataset creation quite a chore.

There are 1400 'Hey Marvin' here for comparison https://drive.google.com/file/d/1dreV5fBIwzdcJnXEueYwc4NeWCyufdS-/view?usp=share_link

I have a tflite that just needs 20ms 16bit 16Khz audio chunks of Hey Marvin on index (0)
https://drive.google.com/file/d/1bGf_b8imzPZJNYDUWR94mWuV0deSMVzM/view index (1) can be used as an inverse VAD if wished
